### ISSUE TITLE
Migrate Ganga to python3

### ISSUE THREAD

**Issue by egede** [ganga-devs/ganga#1485]
We have decided to migrate Ganga to python3. @alexanderrichards has already made very good progress on the Core, Dirac and LHCb packages. The idea is that when these packages are done (which reflect the absolute majority of the use), the shift will be made. There will be no attempt to keep python2 compatibility. The the Ganga 7 series will be terminated as the last one to have python2 compatibility and the Ganga 8 series will only have python3 compatibility.

@JacekHoleczek Any effort available to update the T2K package over the next few weeks? Otherwise it will most likely sit in the code tree of Ganga 8 but fail to load until updated.

**Comment by JacekHoleczek** [comment]
Of course, we will make GangaND280 fully compatible with python 3, when really needed (well, we assume that the majority of its source code is already fine). Do you think that the hight CPU usage which I reported in https://github.com/ganga-devs/ganga/issues/1470 will simply be gone in python 3?

But, I do have some more general questions.

How do you expect people to get "python 3" on the existing SL(C) 6 (default python 2.6.6) and / or CentOS 7 (default python 2.7.5) batch clusters (e.g. PBS and SLURM in my case)?

For the moment, it seems to me that the only way would be to use "(mini)conda" so, will you test "python 3" from "(mini)conda"?

I hope, the "python 3" requirement is only for the main Ganga user interface executable.

Just to be sure ... will you still assume that it is the "python 2.6" which may run on remote batch nodes? So, will all "wrapper scripts" be still python 2.6 compliant?

What will then happen to any jobs which Ganga starts on the "local host" ("Interactive" and "Local" backends)? Will you make sure that the Ganga user interface does not "leak" its own "python 3" settings?

The same concerns ROOT jobs. I think Ganga must not "leak" any of its own "python 3" settings to any ROOT jobs (and this includes, for example, ROOT which appears in the "[File_Associations]" and the "[Mergers]").

**Comment by egede** [comment]
The branch https://github.com/ganga-devs/ganga/tree/python3 is where the python3 development is happening. Feel free to check it out and look at the changes required for GangaND280. GangaCore should work within python3 on that branch.

On systems with cvmfs installed, the access to python3 is easy. My feeling is that most systems will have a python3 installation available. Remember that pythion2.6 is no longer supported. As I understand it, `conda` is just a type of virtual environment, so I do not see any special concerns regarding that. 

For the wrapper scripts we will aim to keep python2.6 compatibility for the moment. Thanks for the reminder about this though as it might affect some other developments that we do at the moment.

Ganga is just a python script and doesn't modify the environment. As such Local jobs, the ROOT merger etc. will inherit the environment. Ganga will start with the shebang `#!/usr/bin/env python3` so what `python` or `python2` does should be independent and within your control.

**Comment by JacekHoleczek** [comment]
I guess I was not clear enough.

It seems to me that, in general, almost always, "python 3" will be provided by some non-standard distribution, like "(mini)conda" (btw. none of the batch clusters that I use provides cvmfs).
That means that, running the Ganga user interface executable will require prior setting / modification of some environment variables (e.g. "PATH", "LD_LIBRARY_PATH", "PYTHONPATH").
However, these modified settings must not be "leaked" to any jobs which Ganga itself starts (neither on "remote" nor "interactive", "local", "ROOT merger" backends). These jobs should always get a perfectly clean environment, like a newly started "login shell".

**Comment by egede** [comment]
I think you were clear enough. My point was that you are asking for something that is difficult/impossible to provide. As any other python program, Ganga is not responsible for setting up the environment that allows python to run.

- At the moment, Ganga is a python2.7 program that should be able to run in any python2.7 environment.
- In the future, Ganga will be a python3 program that should be able to run in any python3 environment.

On a centos7 system I just did the following
- Installed miniconda into my home directory
- Removed the symlink ~/miniconda3/bin/python that pointed to python3
- Edited the shebang in ~/miniconda3/bin/conda to say python3 rather than python.

With those changes python works as python2 completely as expected. The only environment variable that is changed by adding conda is PATH.

**Comment by JacekHoleczek** [comment]
I'm afraid I was again not clear enough.

Let me try it this way ...

In the terminal window in which I want to run Ganga, I first export "PATH", "LD_LIBRARY_PATH", "PYTHONPATH" (I mean whatever is needed to get some non-default "python 3" running) and then I start the "ganga" itself.

However, when the running Ganga "spawns" any "remote", "interactive", "local", "ROOT merger" processes, it must always start them with a "clean" environment, like they were started in a completely new "login shell".
In other words, these "spawned" processes must not "inherit" any environment variables (e.g. "PATH", "LD_LIBRARY_PATH", "PYTHONPATH") which Ganga itself (and its non-default "python 3") got when it was started.

**Comment by egede** [comment]
No, I think you are fairly clear. 

What I say is that the request is unreasonable and unnecessary. I even demonstrated this. This is not the current behaviour of Ganga and the change to python3 doesn't really affect that.

**Comment by JacekHoleczek** [comment]
In this case, I guess I need to disagree.

I tried several different batch clusters which provide different ways to get some non-default "python 3" versions (via "Environment Modules", via "Lmod" modules, via "scl" and via "conda").

In all cases, non-standard "python 3" distributions were compiled with non-standard compilers and they use a whole bunch of non-standard versions of utilities and libraries (which replace the system provided ones).

Of course, a simple test like "`import os; os.system("export")`" immediately shows that the non-standard environment settings are "leaked" to the "subshell".

In general, this is what one really expects.
For example, if one wants to add some "optional components" to "python 3" (and Ganga), which need to be compiled, they should be built with exactly the same compilers, libraries and flags as were used to build the used "python 3" itself.

However, it will be a problem if Ganga does not run user's "jobs' related" executables in a clean "login shell" environment.

Executables which are linked against shared libraries will often misbehave / die if a non-compatible library is "dynamically loaded" at runtime.
Executables and shell scripts which call some utilities or compilers will get "improper" non-standard versions of them (e.g. using non-compatible "cc" / "g++", "gfortran" will lead to severe problems).
In particular, ROOT is known to be very sensitive to this kind of problems.

Note that, it is, of course, possible that the user automatically "enables" / "activates" some non-system-default environment in his / her own "login shell" environment. But then it is the user's responsibility to make sure that all his / her "jobs' related" executables (and scripts) are fully compatible with this setup.

**Comment by mesmith75** [comment]
I am not sure I understand this. All you need is to have the python3 executable (or link to one) in your PATH. This is no different to having a `python2.7` one in PATH for the current ganga version.

I don't really see how that is an issue.

**Comment by JacekHoleczek** [comment]
So, in order to make some tests with python 3.6.5 (from a "Lmod" module on my batch cluster), I tried:
```
module load tools/python/3.6.5
virtualenv ganga-python3
source ganga-python3/bin/activate
pip install https://github.com/ganga-devs/ganga/archive/python3.zip
ganga
```
which then immediately failed:
```
Traceback (most recent call last):
  File "/.../ganga-python3/bin/ganga", line 52, in <module>
    standardSetup()
  File "/.../ganga-python3/bin/ganga", line 47, in standardSetup
    from GangaCore.PACKAGE import standardSetup
ModuleNotFoundError: No module named 'GangaCore'
```
If it helps, there exist two files:
```
/.../ganga-python3/lib/python3.6/site-packages/ganga/GangaCore/PACKAGE.py
/.../ganga-python3/lib/python3.6/site-packages/ganga/GangaCore/__pycache__/PACKAGE.cpython-36.pyc
```

**Comment by mesmith75** [comment]
@alexanderrichards I am getting lots of errors of the form:
```python
Exception in thread Ganga_Worker_4:
Traceback (most recent call last):
  File "/cvmfs/sft.cern.ch/lcg/releases/LCG_92python3/Python/3.6.3/x86_64-slc6-gcc62-opt/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/cvmfs/sft.cern.ch/lcg/releases/LCG_92python3/Python/3.6.3/x86_64-slc6-gcc62-opt/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/afs/cern.ch/user/m/masmith/cmtuser/GANGA/GANGA_HEAD/install/ganga/ganga/GangaCore/Core/GangaThread/WorkerThreads/WorkerThreadPool.py", line 159, in __worker_thread
    self.__queue.task_done()
  File "/cvmfs/sft.cern.ch/lcg/releases/LCG_92python3/Python/3.6.3/x86_64-slc6-gcc62-opt/lib/python3.6/queue.py", line 68, in task_done
    raise ValueError('task_done() called too many times')
ValueError: task_done() called too many times

```

**Comment by alexanderrichards** [comment]
Thanks mark I'll take a look

**Comment by JacekHoleczek** [comment]
I have just tried the newest "python3" branch and I still get the same error as in my previous post above.
Could someone, please, tell me what is wrong.

**Comment by alexanderrichards** [comment]
> ModuleNotFoundError: No module named 'GangaCore'

Hi @JacekHoleczek I assume you mean this error? Looks like your install didn't locate ganga on a default `PYTHONPATH`. You could try manually setting the `PYTHONPATH` but I'd be interested to know why pip didn't install it properly.

What do you get for
```bash
which python
which pip
ls $(dirname `which pip`)/../lib/python3.6/site-packages
```

also instead of pip installing the tarball, do you get any different installing directly form git using
```bash
pip install git+https://github.com/ganga-devs/ganga.git
```

**Comment by alexanderrichards** [comment]
> .../ganga-python3/lib/python3.6/site-packages/ganga/GangaCore/PACKAGE.py

I just noticed that the site-package has lowercase ganga in it. This means that presumably the imports should be:
```python
from ganga.GangaCore.PACKAGE import standardSetup
```

**Comment by alexanderrichards** [comment]
actually this is expected as this would be how you import things from within a default python interpreter.
Let me investigate further

**Comment by mesmith75** [comment]
Something has been commented out in this branch that adds the necessary location to the python path for pip installation (it was put in when we changed the structure and renamed some packages).

**Comment by alexanderrichards** [comment]
ah great, thanks @mesmith75 that was the very next thing I was going to check. Have you added that to the mods_patch branch?

**Comment by alexanderrichards** [comment]
yes indeed:
https://github.com/ganga-devs/ganga/blob/0bdc5d8e0a7bf2f73fb0ae04fd1797b55e7e2235/bin/ganga#L41-L45

I don't remember doing this so presumably it was done automatically as part of the 2to3 conversion?!
Anyway I'll comment it back in and add it to your mods_branch @mesmith75.

We should then merge that branch into python3 so @JacekHoleczek can work with it.

**Comment by alexanderrichards** [comment]
@JacekHoleczek if you want to try it immediately, before we put the fix back into the python3 branch you can install marks mod branch with

```bash
pip install git+https://github.com/ganga-devs/ganga.git@python3_mark_mods3
```

The fix is in there and I checked it works.

**Comment by mesmith75** [comment]
I think that may have been me - I need to comment out those lines for my local development installation on lxplus.

**Comment by alexanderrichards** [comment]
> I think that may have been me - I need to comment out those lines for my local development installation on lxplus.

I assume this is because it was hanging forever? This should be fixed in the commit I put on your mod branch. essentially the filters needed a list() wrapped around them for Python 3

**Comment by JacekHoleczek** [comment]
I am now able to start "ganga-python3", thanks.

Before I start to fiddle with GangaND280, could you, please tell me what are your exact plans for the future?

Are you planning a new 7.1.15 release (and possibly later some more for python2) or do you plan to immediately abandon the 7.1.x series completely and start with 8.0.x for python3?

I am asking it because, as reported in https://github.com/ganga-devs/ganga/issues/1470, I seem to suffer from high CPU usage (I'd really appreciate some help here).
If there are plans for any new 7.1.x releases, it would probably be better to try to fix this problem in the python2 compatible version (I guess it will not go away just by moving to python3).

**Comment by alexanderrichards** [comment]
The python 3 branch will essentially become Ganga 8. I have a little bit more work to do getting the tests passing before I think we are ready to do the switch. Any bugs that are fixed in the meantime will be released on the Ganga 7.1.X branch and I will then port them into python 3 branch. I'm sure you can understand that keeping the two versions in sync like this is painful and I don't intend this to be a long term exercise (a la Python 2/3). As soon as we think the python 3 branch is ready we will release it as Ganga 8 and make the switch. At this point I think we would prefer to fix bugs that come along in the Ganga 8 (Python 3) release. I can't say we definitely won't backport fixes to 7.1.X I guess it would depend on factors like the available manpower and the seriousness of the bug but I would not expect this to happen. We definitely don't want to be in the regime where we are supporting 2 versions simultaneously as we simply don't have the manpower

---
@egede fell free to disagree if you think I have misspoken anywhere :wink:

**Comment by egede** [comment]
@alexanderrichards This is indeed exactly the strategy. For users it should hopefully be a pretty minor issue. to switch from python2 to python3. Given that python2 will cease to be supported by the end of the year, it is a bit of a mute question as well.
